### PR TITLE
Add gpu to aretomo

### PR DESCRIPTION
**Pull request by EuanPyle**
Allows user to specify GPU in the exact same manner as MotionCor2 in rln e.g. user provides a string with colon separated GPU IDs: 0:1:2:3

### CODE REVIEW

File: tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def align_single_tilt_series(
         n_patches_xy: Tuple[int, int],
         alignment_thickness_px: float,
         tilt_angle_offset_correction: bool,
+        gpu_id: str or None,
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py:24]: this type hint should look like 🙂 
```suggestion
        gpu_id: Optional[str],
```

from python 3.10 you can also do
```
str | None
```
see [PEP604](https://peps.python.org/pep-0604/) for details

**Review by alisterburt — COMMENTED** [review]
waiting to see how the API changes at teamtomo/lil_aretomo#26 before moving forward here

**Comment by EuanPyle** [comment]
This now takes str from rln (as 0:1:2 etc), converts to Tuple(int,...) to pass to lil_AT

File: tomography_preprocessing/tilt_series_alignment/aretomo/_utils.py [hunk 2]
```diff
@@ -4,4 +4,5 @@ def gpu_ids_string2tuple(
         gpu_ids: str
     ) -> Tuple:
     """Convert GPU  ID input from colon spaced (1:2:3) to no space inbetween GPU IDs (123)"""
+    gpu_ids = gpu_ids.replace(' ','')
     return tuple(map(int,gpu_ids.replace(':','')))
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/_utils.py:8]:
```suggestion
    return tuple(map(int, gpu_ids.strip().replace(':','')))
```

File: tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py [hunk 3]
```diff
@@ -21,6 +23,7 @@ def align_single_tilt_series(
         n_patches_xy: Tuple[int, int],
         alignment_thickness_px: float,
         tilt_angle_offset_correction: bool,
+        gpu_ids: str or None,
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py:26]:
```suggestion
        gpu_ids: Optional[str],
```

File: tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py [hunk 4]
```diff
@@ -5,6 +5,8 @@
 from rich.console import Console
 from typing import Tuple
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py:6]:
```suggestion
from typing import Optional, Tuple
```

File: tomography_preprocessing/tilt_series_alignment/aretomo/_utils.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+from typing import Tuple
+
+def gpu_ids_string2tuple(
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/_utils.py:3]:
```suggestion
def coerce_gpu_ids(
```

File: tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py [hunk 6]
```diff
@@ -35,6 +38,7 @@ def align_single_tilt_series(
     n_patches_xy: number of patches in x and y for local alignments
     alignment_thickness_px: thickness of intermediate reconstruction during alignments.
     tilt_angle_offset_correction: flag to enable/disable stage tilt offset correction (-TiltCor) in AreTomo
+    gpu_id: string to specify GPUs. GPU identifiers should be separated by colons e.g. 0:1:2:3
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py:41]:
```suggestion
    gpu_ids: string to specify GPUs. GPU identifiers should be separated by colons e.g. 0:1:2:3
```

File: tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py [hunk 7]
```diff
@@ -5,6 +5,8 @@
 from rich.console import Console
 from typing import Tuple
 
+from ._utils import gpu_ids_string2tuple
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py:8]:
```suggestion
from ._utils import coerce_gpu_ids
```

File: tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py [hunk 8]
```diff
@@ -59,6 +63,12 @@ def align_single_tilt_series(
         image_files=tilt_image_df['rlnMicrographName'],
         output_image_file=stack_directory / tilt_series_filename
     )
+    
+    #Convert GPU ID string to tuple
+    gpu_ids = gpu_ids_string2tuple(
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/align_tilt_series.py:68]:
```suggestion
    gpu_ids = coerce_gpu_ids(
```

File: tomography_preprocessing/tilt_series_alignment/aretomo/batch.py [hunk 9]
```diff
@@ -37,6 +38,7 @@ def batch_aretomo(
     alignment_thickness: thickness of intermediate reconstructions during alignments in px.
     tomogram_name: 'rlnTomoName' for a specific tilt-series.
     tilt_angle_offset_correction: flag to enable/disable stage tilt offset correction (-TiltCor) in AreTomo
+    gpu_id: string to specify GPUs. GPU identifiers should be separated by colons e.g. 0:1:2:3
```
Comment (alisterburt) [inline tomography_preprocessing/tilt_series_alignment/aretomo/batch.py:41]:
```suggestion
    gpu_ids: string to specify GPUs. GPU identifiers should be separated by colons e.g. 0:1:2:3
```

**Review by alisterburt — APPROVED** [review]

**Pull request merged**